### PR TITLE
IRGen: Use the header size for the initial offset in non fixed heap l…

### PR DESCRIPTION
**Pull request by aschwaighofer**
…ayouts

When the first element in the heap layout was non fixed we would use the
mininum size of the total heap layout for the initial offset. This would
create unneccessary large heap layouts.

rdar://61716736

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test source compatibility

**Comment by aschwaighofer** [comment]
The source compat failure is unrelated:

``
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.4.sdk/usr/include/stdlib.h:61:10: fatal error: cannot open file '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS13.4.sdk/usr/include/Availability.h': Operation not permitted
#include <Availability.h>
         ^
1 error generated.
``

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test source compat

File: lib/IRGen/GenHeap.cpp [hunk 1]
```diff
@@ -285,7 +285,7 @@ static llvm::Value *calcInitOffset(swift::irgen::IRGenFunction &IGF,
                                    const swift::irgen::HeapLayout &layout) {
   llvm::Value *offset = nullptr;
   if (i == 0) {
-    auto startoffset = layout.getSize();
+    auto startoffset = layout.getHeaderSize();
```
Comment (slavapestov) [inline lib/IRGen/GenHeap.cpp:288]: Can you rename this to startOffset?

**Review by slavapestov — COMMENTED** [review]
I think it's unfortunate that we support `partial_apply` contexts with non-fixed offsets. Did we give up on using SILBoxType for this purpose?

**Comment by aschwaighofer** [comment]
The source compatibility failure is unrelated:

```
SIL memory lifetime failure in @$s10Kingfisher3leeoiySbAA0A15OptionsInfoItemO_ADtF: memory is initialized, but shouldn't
memory location:   %76 = alloc_stack $KingfisherOptionsInfoItem    // users: %87, %81, %77
at instruction:   dealloc_stack %76 : $*KingfisherOptionsInfoItem // id: %81
```

File: lib/IRGen/GenHeap.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/IRGen/GenHeap.cpp:288]: Will change in a follow-up.

**Comment by aschwaighofer** [comment]
I don't think we have given up on using SILBoxType's for the representation of closures. 'Just' needs some work in SILGen, SIL, IRGen.

**Pull request merged**